Feat visibility bsd

samaloney opened the pull request.

nicHoch commented on stixcore/products/level0/quicklook.py:
```diff
@@ -148,15 +148,15 @@ def from_levelb(cls, levelb):
 
         control['compression_scheme_counts_skm'], control['compression_scheme_counts_skm'].meta =\
-            _get_compression_scheme(packets, 'NIX00272')
+            _get_compression_scheme(packets, ['NIXD0101', 'NIXD0102', 'NIXD0103'])
```
i would prefer the "old" approach so the code do not have to know hard coded which are the nix parametrs used for decompression. As this might be false as the actually uncompression takes the nix names from somewhere else.

nicHoch commented on stixcore/products/level0/science.py:
```diff
@@ -266,14 +266,253 @@ def is_datasource_for(cls, *, service_type, service_subtype, ssid, **kwargs):
+
+        real = packets.get_value('NIX00263')
+        real_var = packets.get_value('NIX00263', attr='error')
+        imaginary = packets.get_value('NIX00264')
+        imaginary_var = packets.get_value('NIX00264', attr='error')
+
+        mesh = np.ix_(np.arange(unique_times.size), detector_id[0][0],
```
we should store the skm values used for the decompression:

`control['compression_scheme_vis_skm'], \     control['compression_scheme_vis_skm'].meta = \    _get_compression_scheme(packets, 'NIX00263')`

nicHoch commented on stixcore/products/level0/science.py:
```diff
@@ -266,14 +266,253 @@ def is_datasource_for(cls, *, service_type, service_subtype, ssid, **kwargs):
+        data['triggers'] = triggers
+        data['triggers_err'] = np.sqrt(triggers_var)
+        data['counts'] = full_counts * u.ct
+        data['counts_err'] = np.sqrt(full_counts_var) * u.ct
+        data['control_index'] = 0
+
```
we should store the skm values used for the decompression:

`control['compression_scheme_triggers_skm'], \     control['compression_scheme_vis_skm'].meta = \    _get_compression_scheme(packets, 'NIX00267')`  also for counts

nicHoch commented on stixcore/products/product.py:
```diff
@@ -213,11 +213,11 @@ def from_packets(cls, packets):
         control['compression_scheme_counts_skm'], \
             control['compression_scheme_counts_skm'].meta = \
-            _get_compression_scheme(packets, 'NIX00260')
+            _get_compression_scheme(packets, ['NIXD0007', 'NIXD0008', 'NIXD0009'])
 
         control['compression_scheme_triggers_skm'], \
             control['compression_scheme_triggers_skm'].meta = \
-            _get_compression_scheme(packets, 'NIX00242')
+            _get_compression_scheme(packets, ['NIXD0010', 'NIXD0011', 'NIXD0012'])
 
         control['time_stamp'] = np.array(packets.get_value('NIX00402'))
```
remove that hard coded skm values. as the nixes are not constant and also not all packets have this information (visibilities->counts)

nicHoch reviewed (changes requested):
i looked mainly on the decompression part. and i think this should be reworked.
but the actual Visibilitie and Spectrogram logic looks good

samaloney commented:
I reworked (nearly identical to before) `_get_compression_scheme` and pull the calls out from the common `Control.from_packets` method and added them to the specific products updating the NIX values.

nicHoch reviewed (approved):
looks good now
